src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -795,39 +817,53 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                         int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
                         if (readBytes == 0)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         _internalBufferCount += readBytes;
                         if (_internalBufferCount < SecureChannel.ReadHeaderSize)
                         {
                             await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize).ConfigureAwait(false);
                         }
-                    }
-                    Debug.Assert(_internalBufferCount >= SecureChannel.ReadHeaderSize);
 
-                    // Parse the frame header to determine the payload size (which includes the header size).
-                    int payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
-                    if (payloadBytes < 0)
-                    {
-                        throw new IOException(SR.net_frame_read_size);
-                    }
+                        if (payloadBytes == int.MaxValue)
+                        {
+                            // Parse the frame header to determine the payload size unless already done above.
+//                            payloadBytes = GetFrameSize(_internalBuffer.Span.Slice(_internalOffset));
+                            payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+                            if (payloadBytes < 0)
+                            {
+                                throw new IOException(SR.net_frame_read_size);
+                            }
+                        }
 
-                    // Read in the rest of the payload if we don't have it.
-                    if (_internalBufferCount < payloadBytes)
-                    {
-                        await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        // Read in the rest of the payload if we don't have it.
+                        if (_internalBufferCount < payloadBytes)
+                        {
+                            await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        }
```

Comment:
Actually, I don't understand why any of the code above is inside the while loop. We are only ever going to execute this once, right? Or am I misunderstanding?

Reply:
I'm not sure I fully understand the comment but you can read and process TLS frame that yields no data. From the prospective the frame processing needs to be inside of the loop. Examples may be session tickets, renegotiation or CloseNotification.

Reply:
Ah yeah, I understand. Sorry.